### PR TITLE
DEV: Re-implement commission models to return correct results in the case of multiple fills

### PR DESCRIPTION
**Pull request by jbredeche**

### CODE REVIEW

**Comment by jbredeche** [comment]
(still have a few test failures, working on it)

File: zipline/algorithm.py [hunk 1]
```diff
@@ -56,7 +56,7 @@
     OrderInBeforeTradingStart)
 from zipline.finance.trading import TradingEnvironment
 from zipline.finance.blotter import Blotter
-from zipline.finance.commission import PerShare, PerTrade, PerDollar
+from zipline.finance.commission import PerShare, CommissionModelBase
```
Comment (ssanderson) [inline zipline/algorithm.py]: Most of our other base classes leave out `Base` in the name (e.g. `PipelineLoader`, `TradingControl`, `ExecutionStyle`).

File: zipline/finance/commission.py [hunk 2]
```diff
@@ -12,113 +12,162 @@
...
+        ----------
+        order: the order whose transaction we are processing.  Its `commission`
+            field is up to date with how much commission has been attributed
+            to this order so far.
+
+        transaction: the transaction we are processing.
+
+        Returns
+        -------
+        float: The additional commission we should attribute to this order.
```
Comment (ssanderson) [inline zipline/finance/commission.py]: Worth noting that this is a dollar amount?

File: zipline/finance/commission.py [hunk 3]
```diff
@@ -12,113 +12,162 @@
...
-        means three cents per share, which is a very conservative
-        (quite high) for per share costs.
-        min_trade_cost parameter is the minimum trade cost
-        regardless of the number of shares traded (e.g. $1.00).
-        """
-        self.cost = float(cost)
-        self.min_trade_cost = None if min_trade_cost is None\
-            else float(min_trade_cost)
+        self.cost_per_share = float(cost)
+        self.min_trade_cost = None
```
Comment (ssanderson) [inline zipline/finance/commission.py]: Why not just `self.min_trade_cost = min_trade_cost`.  Not sure what the use-case is for passing a non-float.  (This also means we're actually accepting things like strings here...)

File: zipline/finance/commission.py [hunk 4]
```diff
@@ -12,113 +12,162 @@
...
-    def calculate(self, transaction):
+    def calculate(self, order, transaction):
         """
-        returns a tuple of:
-        (per share commission, total transaction commission)
+        Pay commission based on dollar value of shares.
         """
-        cost_per_share = transaction.price * self.cost
-        return cost_per_share, abs(transaction.amount) * cost_per_share
+        cost_per_share = transaction.price * self.cost_per_dollar
```
Comment (ssanderson) [inline zipline/finance/commission.py]: Curious: does the commission get calculated `before or after slippage has been applied?

File: tests/test_algorithm.py [hunk 5]
```diff
@@ -1555,13 +1558,29 @@ def handle_data(context, data):
                 for val in sublist]
 
             self.assertEqual(len(all_txns), 67)
-            first_txn = all_txns[0]
+            # all_orders are all the incremental versions of the
+            # orders as each new fill comes in.
+            all_orders = [order for orders in
```
Comment (ssanderson) [inline tests/test_algorithm.py]: :confounded:

Comment (ssanderson) [inline tests/test_algorithm.py]: I think this is just `toolz.concat(results['orders'])`:

```
(Pdb++) all_orders == list(toolz.concat(results['orders']))
True
```
(`toolz.concat` is just a more-clearly-named `itertools.chain.from_iterable`).

Comment (jbredeche) [inline tests/test_algorithm.py]: cool, thanks

File: zipline/finance/commission.py (hunk 4 above)
Comment (jbredeche) [inline zipline/finance/commission.py]: Afterwards.  Slippage is applied when we generate the transaction.

File: zipline/finance/commission.py (hunk 3 above)
Comment (jbredeche) [inline zipline/finance/commission.py]: Agreed.

File: zipline/finance/commission.py (hunk 2 above)
Comment (jbredeche) [inline zipline/finance/commission.py]: yup, done

File: zipline/algorithm.py (hunk 1 above)
Comment (jbredeche) [inline zipline/algorithm.py]: cool, changed

**Pull request merged**